### DISCUSSION TITLE
SGR 10..20 (font selection) support

### DISCUSSION THREAD

**Discussion by magistau** [ghostty-org/ghostty#2382] [Feature Requests, Ideas]
[ECMA-48] defines SGR 10..20 for switching between different fonts. Although I'm not aware of any terminals actually supporting this, it would be a nice feature to have.

[ECMA-48]: https://ecma-international.org/wp-content/uploads/ECMA-48_5th_edition_june_1991.pdf

**Comment by mitchellh** [comment]
I'm aware of this SGR from the standard but my understanding is that almost no terminals support it because it isn't clear how grid metrics are supposed to work with changing fonts. 

By "grid metrics", as background: every font defines a unique grid dimension, decoration position (underline, strikethrough), etc. This is unique to the font. If you configure Ghostty (or any other terminal) with one font and then a different one, your grid size (columns x rows) may change, or underline may move higher relative to the cell baseline, etc. 

My belief is SGR 10-20 was designed in an era more rampant with bitmap fonts (as opposed to scalable fonts) which fit into a predefined grid size rather than the grid size adapting to the font. This way, switching fonts is a complete non-issue. But with modern, scalable fonts, I'm not sure how this works. 

I think I have some questions:

* What terminals _do_ support this?
* For terminals that support this, how do they handle grid metrics for differing fonts?
* What programs support this and for what use case?

**Reply by magistau** [reply]
Isn't that problem solved already by the codepoint-map handler?

**Reply by mitchellh** [reply]
Hm, good point, but the expected use case for that is single glyphs moreso than larger runs of text. When using codepoint map (or codepoint fallback in general), we reuse the metrics from the primary font. I wonder if this would look poorly in a larger setting. It is worth experimenting with!

**Reply by magistau** [reply]
> * What terminals _do_ support this?

None that I used myself; [pangoterm] and [mintty] seem to support that. See also microsoft/terminal#6779

> * What programs support this and for what use case?

I don't think there are a lot of programs that support it right now, with that feature being barely supported anywhere. Still, there are some possible use cases:
- [Monaspace] has "mix fonts together" in their philosophy, although that is not actually supported by major editors (see githubnext/monaspace#30)
- artificial scripts like [Tengwar], [Standard Galactic Alphabet], [pIqaD] and [Phyrexian] can not be represented using codepoint-map functionality, because most of them use Basic Latin block for encoding as they don't have any allocated Unicode codepoints
- languages such as [Uiua] require special fonts for comfortable usage (this is what I had in mind when I thought I need an SGR 10..20 implementation)

[pangoterm]: https://leonerds-code.blogspot.com/2011/09/libvtermpangoterm-and-tickit.html
[mintty]: https://github.com/mintty/mintty/wiki/Tips#alternative-fonts
[monaspace]: https://monaspace.githubnext.com
[tengwar]: https://en.wikipedia.org/wiki/Tengwar
[standard galactic alphabet]: https://minecraft.wiki/w/Enchanting_Table#Standard_Galactic_Alphabet
[piqad]: https://en.wikipedia.org/wiki/Klingon_scripts
[phyrexian]: https://mtg.fandom.com/wiki/Phyrexian_(language)
[uiua]: https://www.uiua.org

**Reply by PMunch** [reply]
Adding my two cents here as I've been looking for a terminal with this functionality. For context I got interested in font-based code highlighting after Monaspace was released. I'm working on an e-ink based project where I don't have colour, so using different fonts would be a great alternative.

I made a patch for Vim that was accepted, so now you can highlight things in Vim using these font selectors. I tried it out in pangoterm (after patching it to work properly) and it works well. But it's a bit of a chicken and egg situation with not a lot of programs having support because few terminal emulators supports it because there aren't really any programs using it.

The easiest solution to the question of what to do with different font metrics is probably to just render the text of the new font with its metrics, but placing each glyph in the original grid. So if you specify a font that's too big for example it will clip, and if your main font and any alternate font doesn't agree on letter spacing then the alternate font will look weird. This means that it is up to the user to define a set of alternate fonts that actually match up, which I think is fine for such a niché feature. Since there aren't a lot of programs that make use of this feature I think that anything would be better than nothing in regards to implementation, there's not a lot of programs that _can_ break if things are implemented "incorrectly".
